app/src/test/java/com/example/cleanarchitecture/viewmodel/CharacterViewModelTest.kt

```diff
@@ -0,0 +1,144 @@
+package com.example.cleanarchitecture.viewmodel
+
+import androidx.arch.core.executor.testing.InstantTaskExecutorRule
+import com.example.domain.entity.CharacterData
+import com.example.domain.entity.CharacterDataContainerData
+import com.example.domain.entity.CharacterDataWrapperData
+import com.example.domain.repository.CharacterRepository
+import com.example.domain.usecase.CharacterUseCase
+import com.example.domain.usecase.CharacterUseCaseImpl
+import com.example.domain.utils.Result
+import com.nhaarman.mockitokotlin2.mock
+import com.nhaarman.mockitokotlin2.verify
+import com.nhaarman.mockitokotlin2.whenever
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.test.TestCoroutineDispatcher
+import kotlinx.coroutines.test.resetMain
+import kotlinx.coroutines.test.setMain
+import org.junit.After
+import org.junit.Assert.assertEquals
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.mockito.junit.MockitoJUnitRunner
+
+@ExperimentalCoroutinesApi
+@RunWith(MockitoJUnitRunner::class)
+class CharacterViewModelTest {
+
+    private val testDispatcher = TestCoroutineDispatcher()
+
+    @get:Rule
+    val instantTaskExecutorRule = InstantTaskExecutorRule()
+
+    private lateinit var viewModel: CharacterViewModel
+    private lateinit var useCase: CharacterUseCase
+
+    private val characterRepository: CharacterRepository = mock()
+    private val characterDataWrapperDataValidResult: Result.Success<CharacterDataWrapperData> = mock()
+    private val characterDataWrapperDataInvalidResult: Result.Failure = mock()
+    private val characterDataWrapperData: CharacterDataWrapperData = mock()
+    private val exception: Exception = mock()
+    private val characterDataContainerData: CharacterDataContainerData = mock()
+    private val characters: MutableList<CharacterData> = mock()
+    private val successResponseList: List<Data> =
+        listOf(Data(responseType = Status.LOADING), Data(responseType = Status.SUCCESSFUL, data = characterDataWrapperData))
+    private val emptyListResponseList: List<Data> =
+        listOf(Data(responseType = Status.LOADING), Data(responseType = Status.EMPTY_LIST, error = NO_CHARACTERS_MESSAGE))
+    private val failureResponseList: List<Data> =
+        listOf(Data(responseType = Status.LOADING), Data(responseType = Status.ERROR, error = EXCEPTION_MESSAGE))
+
+    @Before
+    fun `set up`() {
+        Dispatchers.setMain(testDispatcher)
+        useCase = CharacterUseCaseImpl(characterRepository)
+        viewModel = CharacterViewModel(useCase)
+    }
+
+    @After
+    fun `after`() {
+        Dispatchers.resetMain()
+        testDispatcher.cleanupTestCoroutines()
+    }
```

Comment:
if you repeat this in each viewmodel test you can move the functionality to a parent class BaseViewModelTest and extend from it

Reply:
done